File: basics/node-programs/05_error_handling.js
/*
	NODE.JS PROGRAM 5: ERROR HANDLING (TRY-CATCH)
	Concepts:
	- 'try...catch' for catching runtime errors.
	- Creating and 'throwing' custom Error objects.
	- The 'finally' block (runs cleanup regardless).
*/

function divide(a, b) {
	// Custom error check
	if (b === 0) {
		throw new Error("Cannot divide by zero!");
	}
	return a / b;
}

// Success Scenario
try {
	console.log("Success Result: 10 / 2 =", divide(10, 2));
} catch (err) {
	console.log("Caught Error:", err.message);
}

// Error Scenario
console.log("\nError Scenario:");
try {
	const result = divide(10, 0);
	console.log("Result is:", result);
} catch (err) {
	console.log("Caught Error (as expected):", err.message);
} finally {
	console.log("Operation Complete (This runs anyway).");
}

File: basics/go-programs/05_error_handling.go
package main

import (
	"errors"
	"fmt"
)

/*
	GO PROGRAM 5: ERROR HANDLING
	Concepts:
	- Go uses a "return error" approach, not try-catch.
	- Usually, functions return (Result, error).
	- If the second return is not nil, an error occurred.
*/

func divide(a, b float64) (float64, error) {
	// Custom error if denominator is zero
	if b == 0 {
		return 0, errors.New("cannot divide by zero")
	}
	return a / b, nil
}

func main() {
	// Success logic
	result, err := divide(10, 2)
	if err != nil {
		fmt.Printf("Something went wrong: %v\n", err)
	} else {
		fmt.Printf("10 / 2 = %.2f\n", result)
	}
	
	// Error logic
	result2, err2 := divide(10, 0)
	if err2 != nil {
		fmt.Printf("Something went wrong (as expected): %v\n", err2)
	} else {
		fmt.Printf("Result is: %.2f\n", result2)
	}
}
